// jet-popin_table/js/jet-popin-table.js

(function (define) {
        'use strict';

        (function (root, factory) {
                if (typeof define === 'function' && define.amd) {
                        define(factory);
                } else if (typeof exports === 'object') {
                        module.exports = factory();
                } else {
                        root.jetPopinTable = factory();
                }
        }(window, function () {
                return (function () {




                        'use strict';


                        /**
                        * A forEach that will work with a NodeList and generic Arrays
                        * @param {array|NodeList} arr The array to iterate over
                        * @param {function} callback Function that executes for each element. First parameter is element, second is index
                        * @param {object} The context to execute callback with
                        */
                        // function forEach(arr, callback, scope) {
                        //         var i, len = arr.length;
                        //         //console.log(callback);
                        //
                        //         //for (i = 0; i < len; i += 1) {
                        //         //        console.log("looping for =-=-=-=-=");
                        //         callback.call(scope, arr, i);
                        //         //}
                        // }

                        /**
                        * A forEach that will work with a NodeList and generic Arrays
                        * @param {array|NodeList} arr The array to iterate over
                        * @param {function} callback Function that executes for each element. First parameter is element, second is index
                        * @param {object} The context to execute callback with
                        */
                        function forEach(arr, callback, scope) {
                                var i, len = arr.length;
                                for (i = 0; i < len; i += 1) {
                                        callback.call(scope, arr[i], i);
                                }
                        }
                        function isDOMElement(obj) {
                                try {
                                        return obj instanceof HTMLElement;
                                } catch (e) {
                                        // Some browsers don't support using the HTMLElement so some extra
                                        // checks are needed.
                                        return typeof obj === 'object' && obj.nodeType === 1 && typeof obj.style === 'object' && typeof obj.ownerDocument === 'object';
                                }
                        }




                        var events = [
                                'select',
                                'hover'
                        ];




                        var jetPopinTable = function()
                        {


                                function jetPopinTable(container , jsonStruct)
                                {
                                        //EVENT sETUPS
                                        this.handlers = {};
                                        //this.node = container;
                                        this.node = '';
                                        this.container = container;
                                        this.jsonStruct = jsonStruct;
                                        this.cardGrid = "";
                                        this.inlet = "";
                                        this.DropdownContainer = '';
                                        this.doLayout();
                                        this.applyStyle();
                                }
                                jetPopinTable.prototype.applyStyle = function()
                                {
                                        var style = document.createElement('style');
                                        style.type = 'text/css';
                                        style.innerHTML = '.container2 { grid-template-columns: '+this.jsonStruct.settings.columns+'; width : '+this.jsonStruct.settings.width+'; }';
                                        document.getElementsByTagName('head')[0].appendChild(style);
                                }

                                jetPopinTable.prototype.emit=function (instance, name) {
                                        var args = [].slice.call(arguments, 2);
                                        var Textnodes;
                                        console.log(instance.DropdownContainer);
                                        if (events.indexOf(name) > -1) {
                                                console.log("instance==================================");
                                                console.log(instance);
                                                //
                                                console.log("name==================================");
                                                console.log(name);

                                                if (instance.handlers[name] && instance.handlers[name] instanceof Array) {
                                                        console.log("Cocksucker is now working ");
                                                        forEach(instance.handlers[name], function (handle) {
                                                                window.setTimeout(function () {
                                                                        console.log('============handle=====================');
                                                                        console.log(handle);
                                                                        console.log('============Callback=====================');
                                                                        console.log(handle.callback);
                                                                        // console.log('============context=====================');
                                                                        // console.log(handle[0].context);
                                                                        handle.callback.apply(handle.context, args);

                                                                        //        console.log(name);
                                                                        console.log("Cocksucker is now working ");
                                                                        if (name == 'select'){

instance.DropdownContainer.classList.remove("open");
                                                                                //console.log("Cocksucker is now working ");

                                                                        }


                                                                }, 0);
                                                        });
                                                }
                                        } else {
                                                throw new Error(name + ' event cannot be found on jetPopinTable.');
                                        }
                                }.bind(this);
                                jetPopinTable.prototype.doLayout = function ()
                                {

                                        let superbutton = this.drawButton();
                                        let container = this.dropDownMenu();

                                        this.dropDownMenuItems(container);



                                        var click =function (e) {
                                                var parent = (e.target || e.currentTarget).parentNode;
                                                //  console.log(parent);
                                                console.log("In the event thing" );
                                                var data = JSON.parse(parent.getAttribute('clickkey'));
                                                console.log(data);
                                                // var leaves = parent.parentNode.querySelector('.tree-child-leaves');
                                                // if (leaves) {
                                                //         if (leaves.classList.contains('hidden')) {
                                                //                 self.expand(parent, leaves);
                                                //         } else {
                                                //                 self.collapse(parent, leaves);
                                                //         }
                                                // } else {
                                                //         console.log("--------------------------------------------");
                                                //         console.log("slef . parent :: "+ self.parent);
                                                //         console.log(" parent :: "+ parent);
                                                //         console.log("slef :: "+ self);
                                                //         console.log("E.target :: "+ e.target);
                                                //         console.log("--------------------------------------------");
                                                this.emit(this, 'select', {
                                                        target: e,
                                                        data: data,
                                                        parento :parent
                                                });
                                                // }
                                        }.bind(this);

                                        //attack the events
                                        // forEach(self.data, function (item) {
                                        //         //var a;
                                        //         //a++;
                                        //         leaves.push(renderLeaf.call(self, item ,a));
                                        // });
                                        // container.innerHTML = leaves.map(function (leaf) {
                                        //         return leaf.outerHTML;
                                        // }).join('');
                                        //
                                        //
                                        //
                                        //

                                        forEach(container.querySelectorAll('.container2'), function (node) {
                                                node.onclick = click;
                                                //  console.log(node.previousSibling);
                                        });

                                }

                                jetPopinTable.prototype.drawButton = function()
                                {
                                        let baseContainer = document.getElementById(this.container);

                                        // <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown"
                                        // >Tutorials
                                        // <span class="caret"></span></button>
                                        let button = document.createElement("button");
                                        button.classList.add("btn");
                                        button.classList.add("btn-default");
                                        button.classList.add("dropdown-toggle");
                                        button.setAttribute("type","button");
                                        button.setAttribute("data-toggle","dropdown");
                                        button.innerHTML = this.jsonStruct.settings.buttonText;

                                        baseContainer.appendChild(button);

                                        button.onclick = function(e){
                                                var parent = (e.target || e.currentTarget).parentNode;
                                                console.log('=============',parent);

                                                this.DropdownContainer = baseContainer;
                                                //this.DropdownContainer.classList.add("dropdown","open");
                                                if (this.DropdownContainer.classList.contains('open'))
                                                {
                                                        this.DropdownContainer.classList.remove("open");
                                                }else {
                                                        this.DropdownContainer.classList.add("dropdown","open");
                                                }



                                                console.log(baseContainer);
                                        }.bind(this);
                                }

                                //create the drop down container
                                //
                                jetPopinTable.prototype.dropDownMenu = function()
                                {


                                        // <ul class="dropdown-menu">
                                        //
                                        //         <li class = "container2" >
                                        //
                                        //   <div class="grid-item">home</div>
                                        //   <div class="grid-item">22 olvider palace</div>
                                        //   <div class="grid-item">bramley</div>
                                        //
                                        //
                                        //
                                        //         </li>
                                        //   <li class = "container2">
                                        //           <div class="grid-item">home</div>
                                        //           <div class="grid-item">22 olvider palace</div>
                                        //           <div class="grid-item">bramley</div></li>
                                        //   <li class = "container2"><a href="#">JavaScript</a></li>
                                        //   <li class="divider"></li>
                                        //   <li class = "container2"><a href="#">About Us</a></li>
                                        // </ul>


                                        //get the container2
                                        let baseContainer = document.getElementById(this.container)
                                        let ulMenu = document.createElement('ul');
                                        ulMenu.classList.add( 'dropdown-menu');
                                        baseContainer.appendChild(ulMenu);
                                        return ulMenu;




                                }
                                jetPopinTable.prototype.dropDownMenuItems = function(container)
                                {
                                        // get the data list
                                        //
                                        console.log(this.jsonStruct.properties)

                                        let length = this.jsonStruct.properties.length;
                                        let dataArray = this.jsonStruct.properties;
                                        for (let x = 0 ; x< length ;x++)
                                        {

                                                // creat the li tag then the div link
                                                //
                                                //

                                                let li = document.createElement('li');
                                                li.classList.add("container2");
                                                li.setAttribute("clickKey",dataArray[x].id);
                                                container.appendChild(li)
                                                console.log(Object.keys(dataArray[x]));
                                                //now itterate the keus
                                                let arrayKeysCount = Object.keys(dataArray[x]).length;
                                                let names  = Object.keys(dataArray[x])

                                                for (var key in names) {
                                                        console.log(names[key]);
                                                        console.log( dataArray[x][names[key]]);

                                                        let div = document.createElement('div');
                                                        div.classList.add("grid-item");
                                                        div.innerHTML = dataArray[x][names[key]];
                                                        li.appendChild(div);
                                                }

                                                // for(let y = 0 ; y < arrayKeysCount ; y++)
                                                // {
                                                //
                                                //
                                                //
                                                //
                                                //         //creat the div inside the dropdown
                                                //         //
                                                //         //
                                                //         var keyName=names[y];
                                                //         //console.log(names[y]);
                                                //         //console.log(keyName);
                                                //         let div = document.createElement('div');
                                                //         div.classList.add("grid-item");
                                                //         div.innerHTML = dataArray[x].keyName;
                                                //         li.appendChild(div);
                                                //
                                                //
                                                // }

                                                // dataArray.id;
                                                // dataArray.name;
                                                // dataArray.address;
                                                // dataArray.suburb;
                                                // dataArray.owner;



                                        }




                                }







                                jetPopinTable.prototype.on = function (name, callback, scope) {
                                        if (events.indexOf(name) > -1) {

                                                //console.log(this);
                                                if (!this.handlers[name]) {
                                                        this.handlers[name] = [];
                                                }
                                                this.handlers[name].push({
                                                        callback: callback,
                                                        context: scope
                                                });
                                        } else {
                                                throw new Error(name + ' is not supported by jetTreeView.');
                                        }
                                };

                                return jetPopinTable;

                        }();
                        return jetPopinTable;
                }());
        }));
}(window.define));
