// jet-input-form/js/jet-input-form.js
(function (define) {
        'use strict';

        (function (root, factory) {
                if (typeof define === 'function' && define.amd) {
                        define(ResizerPanelLayout,factory);
                } else if (typeof exports === 'object') {
                        module.exports = factory();
                } else {
                        root.jetInputForm = factory();
                }
        }(window, function () {
                return (function () {

                        //console.log("foomar");

                        var jetInputForm = function()
                        {
                                function jetInputForm(jsonInfo,rootElemnt)
                                {
                                        this.jsonInfo = jsonInfo;
                                        //this.jsonsetup = jsonsetup;
                                        this.rootElement = rootElemnt;
                                        jetInputForm.DoLayout(jsonInfo,rootElemnt);
                                        //console.log("foomar");
                                }
                                function createMainPopup()
                                {
                                        //sample code to be used as an example
                                        var sidenav = document.createElement("div");
                                        sidenav.classList.add("sidenav");
                                        sidenav.classList.add("column");
                                        sidenav.classList.add("notification");
                                        sidenav.classList.add("hiding");
                                        sidenav.classList.add("has-text-centered");
                                        sidenav.setAttribute("id","mySidenav");
                                        sidenav.setAttribute("onclick", "ResizerPanelLayout.togglePanelLeft()");
                                        document.documentElement.style.setProperty('--footerHeight', footerHeight);
                                        sidenav.style.display = 'none';
                                        rz.appendChild(sidenav);


                                }
                                jetInputForm.PopupContainer = function()
                                {
                                        // arguments should include a popup name to distinguish againsed different ones
                                        //
                                        var body2 = document.getElementsByTagName("body")[0];
                                        var formPopup = document.createElement("div");
                                        formPopup.setAttribute("id","popup");
                                        formPopup.setAttribute("data-pop","pop-in");
                                        body2.appendChild(formPopup);
                                        return formPopup;
                                }

                                jetInputForm.DialogContainer= function(popup,formName)
                                {

                                        var dialogC = document.createElement("dialog");
                                        dialogC.setAttribute("id","create-repository");
                                        dialogC.setAttribute("data-pop","pop-in");
                                        dialogC.setAttribute("open"," ");
                                        popup.appendChild(dialogC);

                                        //create the header
                                        //
                                        var headerC = document.createElement("header");
                                        headerC.classList.add("dialog-header");
                                        dialogC.appendChild(headerC);

                                        //remeber the form heading comes from JSON
                                        var Myheading = document.createElement("h1");
                                        Myheading.setAttribute("id","__Dialog_create-repository_Create_a_new_repository");
                                        Myheading.innerHTML=formName ;
                                        headerC.appendChild(Myheading);

                                        //creating the close Button
                                        //
                                        var Myclose = document.createElement("a");

                                        Myclose.setAttribute("id","popupclose");
                                        Myclose.classList.add("close");
                                        //svg
                                        //
                                        var MycloseButton = document.createElementNS("http://www.w3.org/2000/svg", "svg");
                                        MycloseButton.setAttribute("aria-hidden","true");
                                        MycloseButton.classList.add("octicon");
                                        MycloseButton.setAttribute("version","1.1");
                                        MycloseButton.setAttribute('viewBox','0 0 12 16');

                                        //path
                                        var MycloseButtonPath = document.createElementNS("http://www.w3.org/2000/svg","path");
                                        MycloseButtonPath.setAttributeNS(null,"d","M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48z");
                                        headerC.appendChild(Myclose);
                                        MycloseButton.appendChild(MycloseButtonPath);
                                        Myclose.appendChild(MycloseButton);

                                        return dialogC;


                                }

                                jetInputForm.createFormContent = function(theFormContainer,jsonInfo)
                                {
                                        //    console.log(jsonFormElements);

                                        //console.log(jsonInfo.formElements);
                                        //console.log(jsonInfo.postAction);
                                        //console.log(jsonInfo.formId);



                                        var formC = document.createElement("form");
                                        //headerC.classList.add("dialog-header");
                                        //
                                        //formC.setAttribute("action",jsonFormAction);
                                        formC.setAttribute("method","post");
                                        //formC.setAttribute("target","_blank");
                                        formC.setAttribute("id",jsonInfo.formId);
                                        formC.setAttribute("autocomplete",'off');
                                        //formC.setAttribute("type","password");
                                        formC.setAttribute("onsubmit",'jetInputForm.xhr(jsonInfo.formId)');

                                        //    formC.setAttribute("onsubmit",'alert("cunts")');
                                        theFormContainer.appendChild(formC);

                                        var fieldsetC = document.createElement("fieldset");
                                        //headerC.classList.add("dialog-header");
                                        formC.appendChild(fieldsetC);

                                        var DialogConTent =document.createElement("div");
                                        DialogConTent.classList.add("dialog-content");
                                        fieldsetC.appendChild(DialogConTent);

                                        // <div class="row-component">
                                        //     <div class="text-box-component">
                                        //         <div class="label-container">
                                        //             <label for="__TextBox_Name">Name</label>
                                        //         </div>
                                        //         <input id="__TextBox_Name" placeholder="repository name" value="" class="">
                                        //     </div>
                                        // </div>

                                        //sending to the textbox this needs an refix for statement
                                        //
                                        //
                                        let a = jsonInfo.formElements.length;
                                        //console.log(a);
                                        for (var i = 0; i < jsonInfo.formElements.length; i++) {
                                                //console.log(jsonFormElements[i]);
                                                if(jsonInfo.formElements[i].type == "textBox")
                                                {
                                                        jetInputForm.formTextbox(DialogConTent,jsonInfo.formElements[i]);
                                                }
                                                if(jsonInfo.formElements[i].type == "dropBox")
                                                {
                                                        //jetInputForm.formTextbox(DialogConTent,);
                                                        jetInputForm.SelectComponent(DialogConTent,jsonInfo.formElements[i]);
                                                }
                                                if(jsonInfo.formElements[i].type == "checkBox")
                                                {
                                                        jetInputForm.formCreateCheck(DialogConTent,jsonInfo.formElements[i]);

                                                }
                                                if(jsonInfo.formElements[i].type == "fileSelect")
                                                {
                                                        jetInputForm.formFileUpload(DialogConTent,jsonInfo.formElements[i]);

                                                }
                                                if(jsonInfo.formElements[i].type == "legend")
                                                {
                                                        jetInputForm.formCreatelegend(DialogConTent,jsonInfo.formElements[i]);

                                                }
                                                if(jsonInfo.formElements[i].type == "textArea")
                                                {
                                                        jetInputForm.formCreateTextArea(DialogConTent,jsonInfo.formElements[i]);

                                                }
                                        }
                                        //jetInputForm.formTextbox(DialogConTent);
                                        //jetInputForm.formCreateCheck(DialogConTent);

                                        //jetInputForm.SelectComponent(DialogConTent);
                                        //jetInputForm.formFileUpload(DialogConTent);
                                        //
                                        return formC;

                                }


                                //-------------------------------------------------
                                //                  FORM COMPONENTS
                                //-------------------------------------------------

                                //defaultOptions
                                //
                                jetInputForm.formFileUpload = function (DialogConTent,jsonFormElements)
                                {
                                        //                        <div class="row-component">
                                        //
                                        //                            <div class="text-box-component">
                                        //
                                        //                                <div class="label-container">
                                        //
                                        //                                    <label for="__TextBox_Local_path">Local path</label>
                                        //                                </div>
                                        //                                <input id="__TextBox_Local_path" placeholder="repository path" value="C:\Users\Jono\Documents\GitHub">
                                        //                            </div>
                                        //                            <button class="button-component" type="button">Choose…</button>
                                        //                        </div>
                                        //
                                        //
                                        //

                                        var rowComponent = document.createElement("div");
                                        rowComponent.classList.add("row-component");
                                        DialogConTent.appendChild(rowComponent);

                                        var textBoxComponent = document.createElement("div");
                                        textBoxComponent.classList.add("text-box-component");
                                        rowComponent.appendChild(textBoxComponent);

                                        var labelContainer= document.createElement("div");
                                        labelContainer.classList.add("label-container");
                                        textBoxComponent.appendChild(labelContainer);

                                        var label = document.createElement("label");
                                        label.innerHTML = jsonFormElements.labelName;
                                        labelContainer.appendChild(label);

                                        var input = document.createElement("input");
                                        input.setAttribute("id",jsonFormElements['id']);
                                        input.setAttribute("name",jsonFormElements['dataName']);
                                        input.setAttribute("autocomplete",'off');
                                        //input.setAttribute("type","password");
                                        input.setAttribute("placeholder","repository path");
                                        textBoxComponent.appendChild(input);

                                        var file_input = document.createElement('input');
                                        file_input.type = 'file';
                                        file_input.style = 'display:none';
                                        file_input.setAttribute("id","ffile");
                                        //file_input.setAttribute("onchange","jetInputForm.selectFile(event.files)")
                                        rowComponent.appendChild(file_input);
                                        //document.getElementById("ffile").addEventListener("change", jetInputForm.selectFile, false);

                                        var button = document.createElement("button");
                                        button.classList.add("button-component");
                                        button.setAttribute("type","button");
                                        //button.setAttribute("onclick","document.getElementById('ffile').click()");
                                        button.setAttribute("multiple",'');
                                        button.setAttribute("id",'foo');
                                        button.innerHTML="choose File";
                                        rowComponent.appendChild(button);
                                        jetInputForm.selectFile();


                                }
                                jetInputForm.selectFile = function()
                                {

                                        //pass the name of the file holder
                                        //read the file contents
                                        var fileupload = document.getElementById("ffile");
                                        var fileupload2 = document.getElementById("file2fileupload");

                                        var filePath=document.getElementById("path32");

                                        var button = document.getElementById("foo");
                                        button.onclick = function () {
                                                fileupload.click();
                                        };
                                        fileupload.onchange = function () {
                                                var fileName = fileupload.value.split('\\')[fileupload.value.split('\\').length - 1];
                                                //console.log(fileName);
                                                fileupload2.value =  fileName;
                                                //    filePath.value =  fileName;
                                        };

                                }

                                //checkbox
                                jetInputForm.formCreateCheck = function(DialogConTent,jsonFormElements)
                                {
                                        //console.log("=================");
                                        //console.log(jsonFormElements);
                                        var rowComponent = document.createElement("div");
                                        rowComponent.classList.add("row-component");
                                        DialogConTent.appendChild(rowComponent);

                                        var checkBoxComponent = document.createElement("div");
                                        checkBoxComponent.classList.add("checkbox-component");
                                        rowComponent.appendChild(checkBoxComponent);

                                        var inputc = document.createElement("input");
                                        inputc.setAttribute("id",jsonFormElements['id']);
                                        inputc.setAttribute("name",jsonFormElements['dataName']);
                                        inputc.setAttribute("type","checkbox");



                                        var label = document.createElement("label");
                                        label.innerHTML = jsonFormElements['labelName'];




                                        checkBoxComponent.appendChild(inputc);
                                        checkBoxComponent.appendChild(label);
                                }
                                //legend label
                                //
                                jetInputForm.formCreatelegend = function(DialogConTent,jsonFormElements)
                                {

                                        var LegendComponent = document.createElement("legend");


                                        LegendComponent.innerHTML = jsonFormElements['labelName'];
                                        DialogConTent.appendChild(LegendComponent);


                                }
                                //legend label
                                //
                                jetInputForm.formCreateTextArea = function(DialogConTent,jsonFormElements)
                                {
                                        //console.log("in the text Aria ");
                                        //==================================
                                        var rowComponent = document.createElement("div");
                                        rowComponent.classList.add("row-component");
                                        DialogConTent.appendChild(rowComponent);

                                        var textBoxComponent = document.createElement("div");
                                        textBoxComponent.classList.add("text-box-component");
                                        rowComponent.appendChild(textBoxComponent);

                                        var labelContainer = document.createElement("div");
                                        labelContainer.classList.add("label-container");
                                        textBoxComponent.appendChild(labelContainer);

                                        var label = document.createElement("label");
                                        label.innerHTML = jsonFormElements['labelName'];
                                        labelContainer.appendChild(label);

//from here
//<textarea rows="4" cols="50">
//At w3schools.com you will learn how to make a website. We offer free tutorials in all web development technologies.
//</textarea>
                                        var textarea = document.createElement("textarea");
                                        textarea.setAttribute("id",jsonFormElements['id']);
                                        textarea.setAttribute("rows",4);
                                        textarea.setAttribute("name",jsonFormElements['dataName']);
                                        textarea.setAttribute("placeholder",jsonFormElements['placeholder']);
                                        textarea.setAttribute("autocomplete",'off');
                                        //inputc.setAttribute("type","lono");

                                        textBoxComponent.appendChild(textarea);
                                        //==================================

                                //        var LegendComponent = document.createElement("legend");


                                //        LegendComponent.innerHTML = jsonFormElements['labelName'];
                                        //DialogConTent.appendChild(LegendComponent);


                                }

                                //SelectComponent
                                jetInputForm.SelectComponent = function(DialogConTent,jsonFormElements)
                                {
                                        // console.log(jsonFormElements);
                                        //
                                        // console.log(jsonFormElements['id']);
                                        // console.log(jsonFormElements['labelName']);
                                        // console.log(jsonFormElements['option']);//: (3) ["none", "something", "something_else"]
                                        // console.log(jsonFormElements['type']);

                                        var rowComponent = document.createElement("div");
                                        rowComponent.classList.add("row-component");
                                        DialogConTent.appendChild(rowComponent);

                                        //<div class="select-component">
                                        var formSelectComponent = document.createElement("div");
                                        formSelectComponent.classList.add("select-component");
                                        rowComponent.appendChild(formSelectComponent);

                                        // labelledby <label for="__Select_Git_ignore">Git ignore</label>

                                        var formSelectComponentLabel = document.createElement("label");
                                        formSelectComponentLabel.innerHTML = jsonFormElements['labelName'];
                                        formSelectComponent.appendChild(formSelectComponentLabel);

                                        //<select id="__Select_Git_ignore">

                                        var formSelect = document.createElement("select");
                                        formSelect.setAttribute("id",jsonFormElements['id']);
                                        formSelect.setAttribute("name",jsonFormElements['dataName']);
                                        formSelectComponent.appendChild(formSelect);

                                        //    <option value="None">None</option>
                                        for  (var i =0;  i < jsonFormElements['option'].length ; i++)
                                        {
                                                jsonFormElements['option'][i];
                                                var selectOption = document.createElement("option");
                                                selectOption.setAttribute("value",jsonFormElements['option'][i]);
                                                selectOption.innerHTML= jsonFormElements['option'][i];;
                                                formSelect.appendChild(selectOption);
                                        }
                                        // var selectOption = document.createElement("option");
                                        // selectOption.setAttribute("value","none");
                                        // selectOption.innerHTML= "none";
                                        // formSelect.appendChild(selectOption);
                                        //
                                        // var selectOption1 = document.createElement("option");
                                        // selectOption1.setAttribute("value","something");
                                        // selectOption1.innerHTML= "something";
                                        // formSelect.appendChild(selectOption1);






                                }

                                //TextBox Imput
                                jetInputForm.formTextbox = function (DialogConTent,jsonFormElements)
                                {
                                        //         console.log(jsonFormElements);
                                        //         console.log("------------------");
                                        //
                                        // //if (typeof jsonFormElements === 'object')
                                        // //{
                                        //         console.log("is an object");
                                        //         console.log(jsonFormElements['labelName']);
                                        //         console.log(jsonFormElements['id']);
                                        //         console.log(jsonFormElements['placeholder']);
                                        //}

                                        var rowComponent = document.createElement("div");
                                        rowComponent.classList.add("row-component");
                                        DialogConTent.appendChild(rowComponent);

                                        var textBoxComponent = document.createElement("div");
                                        textBoxComponent.classList.add("text-box-component");
                                        rowComponent.appendChild(textBoxComponent);

                                        var labelContainer = document.createElement("div");
                                        labelContainer.classList.add("label-container");
                                        textBoxComponent.appendChild(labelContainer);

                                        var label = document.createElement("label");
                                        label.innerHTML = jsonFormElements['labelName'];
                                        labelContainer.appendChild(label);

                                        var inputc = document.createElement("input");
                                        inputc.setAttribute("id",jsonFormElements['id']);
                                        inputc.setAttribute("name",jsonFormElements['dataName']);
                                        inputc.setAttribute("placeholder",jsonFormElements['placeholder']);
                                        inputc.setAttribute("autocomplete",'off');
                                        //inputc.setAttribute("type","lono");

                                        textBoxComponent.appendChild(inputc);
                                }
                                jetInputForm.footer = function(DialogConTent)
                                {
                                        // <div class="dialog-footer">
                                        //                     <div class="button-group">
                                        //                         <button class="button-component" disabled="" type="submit">Create repository</button>
                                        //                         <button class="button-component" type="button">Cancel</button>
                                        //                     </div>
                                        //                 </div>
                                        //
                                        //
                                        var dialogFooter = document.createElement("div");
                                        dialogFooter.classList.add("dialog-footer");
                                        DialogConTent.appendChild(dialogFooter);

                                        var buttonGroup = document.createElement("div");
                                        buttonGroup.classList.add("button-group");
                                        dialogFooter.appendChild(buttonGroup);

                                        var button1 = document.createElement("button");
                                        button1.classList.add("button-component");
                                        button1.setAttribute("type","submit");
                                        button1.innerHTML = "submit";
                                        buttonGroup.appendChild(button1);

                                        var button2 = document.createElement("button");
                                        button2.classList.add("button-component");
                                        button2.setAttribute("type","button");
                                        button2.innerHTML = "cancel";
                                        buttonGroup.appendChild(button2);

                                }

                                jetInputForm.AllocateInlineContent = function(appendPanel)
                                {
                                        // console.log(appendPanel);
                                        var pane = document.getElementById(appendPanel);
                                        return pane;
                                }

                                jetInputForm.DoLayout = function(jsonInfo,appendPanel)
                                {

                                        // console.log("=============================");
                                        // console.log(jsonInfo.postAction);
                                        //
                                        // console.log("=============================");
                                        //console.log(jsonInfo.formName);
                                        //    var popup = jetInputForm.PopupContainer();
                                        var popup = jetInputForm.AllocateInlineContent(appendPanel);
                                        //    var theFormContainer = jetInputForm.DialogContainer(popup,jsonInfo.formName)
                                        var theform = jetInputForm.createFormContent(popup,jsonInfo);
                                        //jetInputForm.footer(theFormContainer);
                                        jetInputForm.footer(theform);
                                        //console.log("dolayout Initialised !!!1");


                                }





                                jetInputForm.xhr = function functionName(formId) {
                                        //prevent page from reloading
                                        event.preventDefault();

                                        // console.log("in the xhr function" );
                                        // console.log("======================================");
                                        // console.log("======================================");
                                        // console.log("======================================");
                                        // console.log("======================================");

                                        var form = document.getElementById(formId);
                                        var n = document.getElementById(formId).elements;
                                        var data = new FormData(form);




                                        // the function herunder does work but  is primitive I could use formdata but is not supportred by all browsers

                                        function serializeArray(form) {
                                                var field, l, s = [],t ={};
                                                if (typeof form == 'object' && form.nodeName == "FORM") {
                                                        var len = form.elements.length;
                                                        for (var i=0; i<len; i++) {
                                                                field = form.elements[i];
                                                                if (field.name && !field.disabled && field.type != 'file' && field.type != 'reset' && field.type != 'submit' && field.type != 'button') {
                                                                        if (field.type == 'select-multiple') {
                                                                                l = form.elements[i].options.length;
                                                                                for (j=0; j<l; j++) {
                                                                                        if(field.options[j].selected)
                                                                                        t[field.name] = field.options[j].value ;
                                                                                        //    s[s.length] = { key: field.name, value: field.options[j].value };
                                                                                }
                                                                        } else if ((field.type != 'checkbox' && field.type != 'radio') || field.checked) {
                                                                                //    s[s.length] = { key: field.name, value: field.value };
                                                                                t[field.name] =field.value;
                                                                        }
                                                                }
                                                        }
                                                }
                                                //return s;
                                                //
                                                return t;
                                        }

                                        var energy = serializeArray(form);
                                        //
                                        //var postType = "JSON";
                                        var postType = "WWW_FORM";
                                        let who = JSON.stringify(energy);
                                        // console.log(energy);


                                        //    console.log(energy);
                                        loadDoc(who,postType);


                                        function loadDoc(energy,postType)
                                        {
                                                //console.log("energy");
                                                //console.log(postType);
                                                var xhttp = new XMLHttpRequest();
                                                xhttp.onreadystatechange = function() {
                                                        if (this.readyState == 4 && this.status == 200) {
                                                                document.getElementById("paragraph").innerHTML = this.responseText;
                                                        }
                                                };

                                                if (postType== "JSON" )
                                                {
                                                        xhttp.open("POST", 'http://localhost/05_API_tests/post.php', true);
                                                        xhttp.setRequestHeader("Content-type", "application/json;charset=UTF-8");
                                                        xhttp.send(energy);

                                                }

                                                if(postType== "WWW_FORM" )
                                                {
                                                        xhttp.open("POST", 'http://localhost/05_API_tests/post.php', true);
                                                        xhttp.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
                                                        xhttp.send("JSONPostString="+energy+"");
                                                }

                                        }}
                                        return jetInputForm;
                                }();

                                return jetInputForm;



                        }());
                }));
        }(window.define));

        // let xhr = [], i;
        // for(i = 0; i < index.length; i++){ //for loop
        //     (function(i){
        //         xhr[i] = new XMLHttpRequest();
        //         let url = 'http://localhost:7890/1.1/statuses/user_timeline.json?count='+TweetCount+'&screen_name='+index[i]+'&tweet_mode=extended' ;
        //         xhr[i].open("GET", url, true);
        //         xhr[i].onreadystatechange = function(){
        //
        //             if (xhr[i].readyState < 4 )
        //             {
        //
        //                 document.getElementById("Header").innerHTML = "Loading Xhr Request";
        //
        //             }
        //             else if (xhr[i].readyState === 4 && xhr[i].status === 200)
        //             {
        //                 data[i] = JSON.parse(xhr[i].responseText);
        //                 setTimeout(function(){
        //                     //console.log(data[i]);
        //                     //ensure the loop has finished and data is returned in order as the requests
        //                     (i  ===  2) ?document.getElementById("Header").innerHTML = "Request Xhr Response" : "";
        //                     (i  ===  2) ?  prepData(null, data[0]) + prepData(null, data[1]) + prepData(null, data[2]): " ";
        //
        //                 },3000);
        //
        //             }else
        //             {
        //                 callback(status, xhr[i].response);
        //             }
        //         };
        //         xhr[i].send();
        //     })(i);
        //
        // }
