// jet-wizard-datastore/js/jet-wizard-datastore.js

(function (define) {
        'use strict';

        (function (root, factory) {
                if (typeof define === 'function' && define.amd) {
                        define(factory);
                } else if (typeof exports === 'object') {
                        module.exports = factory();
                } else {
                        root.jetWizardDatastore = factory();
                }
        }(window, function () {
                return (function () {




                        var jetWizardDatastore = function()
                        {

                                function jetWizardDatastore()
                                {
                                        this.DataObject = '';
                                        this.Initiate();
                                }

                                jetWizardDatastore.prototype.Initiate = function()
                                {

                                        this.DataObject = {};
                                }

                                jetWizardDatastore.prototype.setKeyValuePair = function(key,value)
                                {
                                        this.DataObject[key] = value;

                                }

                                jetWizardDatastore.prototype.getKeyValuePair = function(key)
                                {
                                        return this.DataObject[key];
                                }
                                jetWizardDatastore.prototype.getAllKeys = function()
                                {
                                        return this.DataObject;
                                }

                                jetWizardDatastore.prototype.clear = function()
                                {
                                        this.DataObject = {};
                                }

                                return jetWizardDatastore;



                        }();

                        return jetWizardDatastore;
                }());
        }));
}(window.define));
